### PR TITLE
Itemized PayPal cart

### PR DESCRIPTION
**Pull request by penelopec**
The changes are all related to the Itemized PayPal cart and are associated with the PayPal pull request:
https://github.com/indico/indico-plugins/pull/69

### CODE REVIEW

File: indico/modules/events/registration/fields/base.py [hunk 1]
```diff
@@ -22,7 +22,6 @@
 
 from indico.modules.events.registration.models.registrations import RegistrationData
 
-
```
Comment (ThiefMaster) [inline indico/modules/events/registration/fields/base.py]: this empty line should remain

File: indico/modules/events/registration/fields/base.py [hunk 2]
```diff
@@ -37,7 +36,7 @@ class RegistrationFormFieldBase(object):
     #: the validator to use when the field is not required
     not_required_validator = Optional
     #: the data fields that need to be versioned
-    versioned_data_fields = frozenset({'is_billable', 'price'})
+    versioned_data_fields = frozenset({'is_billable', 'price','price_info'})
```
Comment (ThiefMaster) [inline indico/modules/events/registration/fields/base.py]: space after the `,`

File: indico/modules/events/registration/fields/base.py [hunk 3]
```diff
@@ -172,6 +171,7 @@ def process_field_data(cls, data, old_data=None, old_versioned_data=None):
         # we don't have field-level billing data here
         del versioned_data['is_billable']
         del versioned_data['price']
+        #del versioned_data['price_info']
```
Comment (ThiefMaster) [inline indico/modules/events/registration/fields/base.py]: shouldn't be there (or should it? in that case don't comment it out ;))

File: indico/modules/events/registration/forms.py [hunk 4]
```diff
@@ -51,9 +51,8 @@ def _check_if_payment_required(form, field):
     if not is_feature_enabled(form.event, 'payment'):
         raise ValidationError(_('You have to enable the payment feature in order to set a registration fee.'))
 
-
```
Comment (ThiefMaster) [inline indico/modules/events/registration/forms.py]: again, please keep the two empty lines between top-level classes/functions

File: indico/modules/events/registration/forms.py [hunk 5]
```diff
@@ -51,9 +51,8 @@ def _check_if_payment_required(form, field):
...
 class RegistrationFormForm(IndicoForm):
-    _price_fields = ('currency', 'base_price')
+    _price_fields = ('currency', 'base_price',  'base_price_info')
```
Comment (ThiefMaster) [inline indico/modules/events/registration/forms.py]: there's an extra space after the `,`

File: indico/modules/events/registration/forms.py [hunk 6]
```diff
@@ -89,6 +88,8 @@ class RegistrationFormForm(IndicoForm):
                               _check_if_payment_required], filters=[lambda x: x if x is not None else 0],
                               widget=NumberInput(step='0.01'),
                               description=_("A fixed fee all users have to pay when registering."))
+    base_price_info = StringField(_("Registration fee info"),
+                               description=_("Used by the PayPal itemized cart."))
```
Comment (ThiefMaster) [inline indico/modules/events/registration/forms.py]: not aligned correctly (see the other fields as an example)

File: indico/modules/events/registration/models/forms.py [hunk 7]
```diff
@@ -164,6 +164,11 @@ class RegistrationForm(db.Model):
         nullable=False,
         default=0
     )
+    #: Base registration fee information (used by PayPal)
```
Comment (ThiefMaster) [inline indico/modules/events/registration/models/forms.py]: let's not mention paypal here and instead also show  it in the normal UI somewhere (tooltip on "registration fee" maybe)

File: indico/modules/events/registration/models/forms.py [hunk 8]
```diff
@@ -164,6 +164,11 @@ class RegistrationForm(db.Model):
...
+    base_price_info = db.Column(
+        db.String, 
+        nullable=True
```
Comment (ThiefMaster) [inline indico/modules/events/registration/models/forms.py]:
```
nullable=False,
default=''
```

File: indico/modules/events/registration/models/registrations.py [hunk 9]
```diff
@@ -147,6 +147,11 @@ class Registration(db.Model):
         db.String,
         nullable=False
     )
+    #: Base registration fee information (used by PayPal)
```
Comment (ThiefMaster) [inline indico/modules/events/registration/models/registrations.py]: see my comments in the other model

**Comment by penelopec** [comment]
Hello Adrian,

I did run the indico db migrate -m 'Add base_prince_info' command to create the new column, I will add this in the comments of the pull.

I’ll fix the indentation in all the files.  Sorry for this, my bad.

For the following, I am not sure on what to do, as when I uncomment the line the code does not work, I get errors.  I will try to check a bit more and perhaps ask your help about these errors.

In indico/modules/events/registration/fields/base.py<https://urldefense.proofpoint.com/v2/url?u=https-3A__github.com_indico_indico_pull_3653-23discussion-5Fr231173268&d=DwMCaQ&c=gRgGjJ3BkIsb5y6s49QqsA&r=PtYen6pomi4Ar8w31gt3-vFlwe-Apq0cAEgTWbu5684&m=kZiX0efowhgUW1T9hr-HsYmK_wPO8PQBggmIaeSxd9s&s=JNhGKAQxhtT_djV-ozHREx0Mijfx7AmoCdC6TeuCnDM&e=>:

> @@ -172,6 +171,7 @@ def process_field_data(cls, data, old_data=None, old_versioned_data=None):

         # we don't have field-level billing data here

         del versioned_data['is_billable']

         del versioned_data['price']

+        #del versioned_data['price_info']

shouldn't be there (or should it? in that case don't comment it out ;))
Thank you!
Penelope

**Comment by penelopec** [comment]
Hello Adrian,

I have made the changes on my cloned repositories (under the same branch).  Do I need to create a new pull request?  I am not familiar with GitHub and most probably I have messed a few things.

Thank you
Penelope

**Comment by ThiefMaster** [comment]
No, as long as you pushed the files to the same branch the PR automatically updated. You can check this here: https://github.com/indico/indico/pull/3653/files

PS: I see you uploaded all files manually via the github website. It'd actually be easier if you committed the changes locally using git and then pushed to your branch.

File: indico/modules/events/registration/models/registrations.py [hunk 10]
```diff
@@ -147,6 +147,12 @@ class Registration(db.Model):
         db.String,
         nullable=False
     )
+    #: Base registration fee information
+    base_price_info = db.Column(
+        db.String,
+        nullable=False
```
Comment (ThiefMaster) [inline indico/modules/events/registration/models/registrations.py]: syntax error due to the missing `,` at the end of this line

File: indico/modules/events/registration/client/js/form/field.js [hunk 11]
```diff
@@ -675,6 +678,14 @@ ndRegForm.directive('ndRadioField', function(url) {
...
                     defaultVal: false,
                     edittype: 'bool_select'
+                }, {
+                    name: 'priceInfo',
+                    index: 'priceInfo',
+                    align: 'center',
+                    width: 50,
+                    editable: true,
+                    edittype: "text",
+                    editoptions: {size: "30", }
```
Comment (mic4ael) [inline indico/modules/events/registration/client/js/form/field.js]: You can remove trailing comma and space.

File: indico/modules/events/registration/client/js/form/tpls/fields/dialogs/billable.tpl.html [hunk 12]
```diff
@@ -19,4 +19,15 @@
         class="info-helper"
         title="{{ 'This price will be multiplied times the number entered by the user' | i18n }}">
     </i>
+    <br>
+    <label class="regFormDialogCaption">
+        {{ 'Price info' | i18n }}
+    </label>
+    <input type="text" name="priceInfo"
+        ng-model="formData.priceInfo"
```
Comment (mic4ael) [inline indico/modules/events/registration/client/js/form/tpls/fields/dialogs/billable.tpl.html]: You could always align those attributes to be in the same column as `type`.

File: indico/modules/events/registration/models/forms.py [hunk 13]
```diff
@@ -164,6 +164,12 @@ class RegistrationForm(db.Model):
         nullable=False,
         default=0
     )
+    #: Base registration fee information
+    base_price_info = db.Column(
+        db.String, 
+        nullable=False,
+        default=''
+)
```
Comment (mic4ael) [inline indico/modules/events/registration/models/forms.py]: This bracket is incorrectly aligned.

File: indico/migrations/versions/20181102_1354_5961a6ace73b_add_base_price_info_column.py [hunk 14]
```diff
@@ -0,0 +1,24 @@
...
+
+# revision identifiers, used by Alembic.
+revision = '5961a6ace73b'
+down_revision = '7e03b2262e9e'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    op.add_column('registrations', sa.Column('base_price_info', sa.String(), nullable=True), schema='event_registration')
```
Comment (ThiefMaster) [inline indico/migrations/versions/20181102_1354_5961a6ace73b_add_base_price_info_column.py]: `nullable=False, server_default=''`, and then use `alter_column()` in the next line to remove the server_default again by setting it to `None`.

you can check existing revisions for examples.

File: indico/modules/events/registration/forms.py [hunk 15]
```diff
@@ -80,6 +80,8 @@ class RegistrationFormForm(IndicoForm):
                               _check_if_payment_required], filters=[lambda x: x if x is not None else 0],
                               widget=NumberInput(step='0.01'),
                               description=_("A fixed fee all users have to pay when registering."))
+    base_price_info = StringField(_("Registration fee info"),
+                                  description=_("Used by the PayPal itemized cart."))
```
Comment (mic4ael) [inline indico/modules/events/registration/forms.py:84]: You can have the declaration of `base_price_info` in a single line

**Comment by mic4ael** [comment]
There is 50 commits which could be easily squashed. This will make the whole PR a bit cleaner.

**Comment by penelopec** [comment]
This request and the indico-plugins: Itemized pay pal cart #82 have been rebased to 2.2-maintenance corresponding branches.  It would be nice if they can be part of indico.

**Comment by ThiefMaster** [comment]
Thanks a lot for this contribution, but currently we don't think this should be part of the Indico core or the base plugins we maintain. The changes add a lot of complexity, and for some parts it's not clear whether it's needed at all:

- [core] price info: Why is this needed, doesn't the field name already serve the purpose?
- [core] base_price_info: Is there any need for this, instead of just using "Registration Fee"?
- [plugin] There's lots of code involving different field types there, which would likely required changes whenever we change something regarding fields in the core

So I think it may make sense in a future Indico version to officially support a way to let the Indico core pass a list of "invoice items" (with titles and prices) to payment plugins (or expose an API to get such a list) where each field implementation is responsible for returning the data in a suitable format. But for now this is not something we want to add so I'm going to close the PRs.

**Pull request closed without merging**